### PR TITLE
Parameterize certs and port for prometheus-metrics

### PR DESCRIPTION
**Pull request by chaitanyaenr**
- By default prometheus-metrics tool uses cert, key located at
  /etc/origin/master on the master node of OpenShift cluster, fetches
  the metrics exposed by prometheus endpoint running on port 8443. This
  commit will enable us to ping endpoints running on different ports,
  which need different certs.
- User can pass the certs, port as arguments while registering the tool
  like pbench-register-tool --name=prometheus-metrics -- --inventory=
  /path/to/inventory --port <port> --cert=/path/to/cert --key=/path/to/key
- prometheus-metrics-tool checks if go is installed, if yes it won't
  check if GOPATH is set. It sets the GOPATH only when go is not
  installed. This commit fixes that.
- Instead of modifying user's .bashrc, this will set the GOPATH, adds
  the binaries to the PATH variable only while running the tool
  specific scripts, commands.
- Upstream prom2json is broken from July 10th 2017, so it might not be
  possible to test this commit.

### CODE REVIEW

File: agent/tool-scripts/datalog/prometheus-metrics-datalog [hunk 1]
```diff
@@ -7,12 +7,15 @@ import os,subprocess,warnings,logging
 interval = int(sys.argv[2])
 master = sys.argv[1]
 tool_log = sys.argv[3]
+port = int(sys.argv[4])
+cert = sys.argv[5]
+key = sys.argv[6]
 os.system('scp "%s:/etc/origin/master/admin.crt" "/run/pbench/keys/admin.crt"' % (master) )
```
Comment (chaitanyaenr) [inline agent/tool-scripts/datalog/prometheus-metrics-datalog]: have to move these commands to the tool-script

File: agent/tool-scripts/prometheus-metrics [hunk 2]
```diff
@@ -145,7 +174,20 @@ case "$mode" in
...
+		if [ -z "$cert" ]; then
+			if [ ! -s /run/pbench/keys/admin.crt ]; then
+				scp $master:/etc/origin/master/admin.crt /run/pbench/keys/admin.crt
+			fi
+			cert=/run/pbench/keys/admin.crt
+		fi
+		if [ -z "$key" ]; then
+			if [ ! -s /run/pbench/keys/admin.key ]; then
+				scp $master:/etc/origin/master/admin.key /run/pbench/keys/admin.key
+		fi
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: Indentation

File: agent/tool-scripts/prometheus-metrics [hunk 3]
```diff
@@ -145,7 +174,20 @@ case "$mode" in
...
+			key=/run/pbench/keys/admin.key
+		fi
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: What happens if the scp's fail?

File: agent/tool-scripts/prometheus-metrics (hunk 2 above)
Comment (chaitanyaenr) [inline agent/tool-scripts/prometheus-metrics]: Fixed

File: agent/tool-scripts/prometheus-metrics (hunk 3 above)
Comment (chaitanyaenr) [inline agent/tool-scripts/prometheus-metrics]: the failure gets logged and the tool exits.

File: agent/tool-scripts/prometheus-metrics [hunk 4]
```diff
@@ -145,7 +174,28 @@ case "$mode" in
...
+					error_log "[$script_name]scp command failed to copy the files from $master, please check if the cert exist in /etc/origin/master directory"
+					exit 1
+				fi
+			fi
+			cert=/run/pbench/keys/admin.crt
+		fi
+		if [ -z "$key" ]; then
+			if [ ! -s /run/pbench/keys/admin.key ]; then
+				scp $master:/etc/origin/master/admin.key /run/pbench/keys/admin.key
+				if [[ $(echo $?) != 0 ]]; then
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: This should be changed too

File: agent/tool-scripts/prometheus-metrics [hunk 5]
```diff
@@ -145,7 +174,28 @@ case "$mode" in
 	start)
 	mkdir -p $tool_output_dir/json
 	while read -u 9 master; do
-		tool_cmd="$script_path/datalog/$tool-datalog $master $interval $tool_log"
+		# fetch keys from openshift master and set default values for cert and key vars if not defined
+		if [ -z "$cert" ]; then
+			if [ ! -s /run/pbench/keys/admin.crt ]; then
+				scp $master:/etc/origin/master/admin.crt /run/pbench/keys/admin.crt
+				if [[ $? != 0 ]]; then
+					error_log "[$script_name]scp command failed to copy the files from $master, please check if the cert exist in /etc/origin/master directory"
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: ... exists...

File: agent/tool-scripts/prometheus-metrics [hunk 6]
```diff
@@ -145,7 +174,28 @@ case "$mode" in
...
+					exit 1
+				fi
+			fi
+			cert=/run/pbench/keys/admin.crt
+		fi
+		if [ -z "$key" ]; then
+			if [ ! -s /run/pbench/keys/admin.key ]; then
+				scp $master:/etc/origin/master/admin.key /run/pbench/keys/admin.key
+				if [[ $(echo $?) != 0 ]]; then
+					error_log "[$script_name]scp command failed to copy the files from $master, please check if the key exist in /etc/origin/master directory"
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: ... exists...

**Comment by chaitanyaenr** [comment]
@ndokos PTAL.

File: agent/tool-scripts/prometheus-metrics (hunk 4 above)
Comment (chaitanyaenr) [inline agent/tool-scripts/prometheus-metrics]: Fixed.

Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: I meant the $(echo $?).

Comment (chaitanyaenr) [inline agent/tool-scripts/prometheus-metrics]: Fixed.

File: agent/tool-scripts/prometheus-metrics [hunk 7]
```diff
@@ -126,12 +151,16 @@ elif [[ ! -s "$inventory" ]] || [[ ! -f "$inventory" ]]; then
...
 ## parse openshift inventory file
 cat "$inventory" | awk -F " " '{ print $1 }' | sed -n '/^\[masters\]/,/^$/p' | awk '{if (NR!=1) {print}}' | sed '/^$/d' > /run/pbench/inv_hosts
 case "$mode" in
 	install)
 	which go &>/dev/null
-	if [[ $(echo $?) != 0 ]]; then
+	if [[ $? != 0 ]]; then
 		yum -y install go &>/dev/null
 		mkdir $pbench_install_dir/.go	
 		echo "GOPATH=$pbench_install_dir/.go" >> ~/.bashrc
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: Do *not* muck around with ~/.bashrc

File: agent/tool-scripts/prometheus-metrics [hunk 8]
```diff
@@ -145,7 +174,28 @@ case "$mode" in
...
+			if [ ! -s /run/pbench/keys/admin.key ]; then
+				scp $master:/etc/origin/master/admin.key /run/pbench/keys/admin.key
+				if [[ $? != 0 ]]; then
+					error_log "[$script_name]scp command failed to copy the files from $master, please check if the key exists in /etc/origin/master directory"
+					exit 1
+				fi	
+			fi
+			key=/run/pbench/keys/admin.key
+		fi
+		tool_cmd="$script_path/datalog/$tool-datalog $master $interval $tool_log $port $cert $key"
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: You need to modify the environment for the tool command, right?

      GOPATH=$pbench_install_dir/.go PATH=$GOPATH/bin:$PATH $script_path/datalog/$tool_datalog ...

And you may also need to "eval $tool_cmd" on line 201 (but I'm not sure about that).

Comment (chaitanyaenr) [inline agent/tool-scripts/prometheus-metrics]: @ndokos GOPATH stuff is tracked as a seperate PR - https://github.com/distributed-system-analysis/pbench/pull/610

File: agent/tool-scripts/prometheus-metrics (hunk 7 above)
Comment (chaitanyaenr) [inline agent/tool-scripts/prometheus-metrics]: @ndokos GOPATH stuff is tracked as a seperate PR - https://github.com/distributed-system-analysis/pbench/pull/610

**Comment by chaitanyaenr** [comment]
Fixes #595 #596

File: agent/tool-scripts/prometheus-metrics [hunk 9]
```diff
@@ -162,14 +162,9 @@ case "$mode" in
 	which go &>/dev/null
 	if [[ $? != 0 ]]; then
 		yum -y install go &>/dev/null
-		mkdir $pbench_install_dir/.go
```
Comment (ndokos) [inline agent/tool-scripts/prometheus-metrics]: Don't you need the mkdir?

**Pull request merged**